src/librbd/crypto/CryptoObjectDispatch.cc

```diff
@@ -118,7 +118,7 @@ struct C_UnalignedObjectReadRequest : public Context {
 
       // send the aligned read back to get decrypted
       req = io::ObjectDispatchSpec::create_read(
-              image_ctx, io::OBJECT_DISPATCH_LAYER_NONE, object_no,
+              image_ctx, io::OBJECT_DISPATCH_LAYER_CACHE, object_no,
```

Comment:
The deadlock was cached (for example) when doing a compare-and-write, which triggered asynchronous write back to the write-around cache. The spawned sub-request was blocked by the overlapping original write.

Reply:
Ack. You could either add the `LIBRADOS_OP_FLAG_FADVISE_FUA` op flag to the sub-request to bypass the cache or create that new helper to avoid hard-coding to `LAYER_CACHE` here (just seems fragile to me),

Reply:
I've implemented your suggestion of get_previous_layer. This solves it.